core/sdk-core/src/main/java/software/amazon/awssdk/core/internal/http/pipeline/stages/AsyncRetryableStage.java

```diff
@@ -122,65 +112,75 @@ private RetryExecutor(SdkHttpFullRequest request, RequestExecutionContext contex
         public CompletableFuture<Response<OutputT>> execute() throws Exception {
             CompletableFuture<Response<OutputT>> future = new CompletableFuture<>();
 
-            long apiCallTimeoutInMillis = getApiCallTimeoutInMillis(context.requestConfig());
-            TimeoutTracker timeoutTracker = timeCompletableFuture(future,
-                                                                  scheduledExecutor,
-                                                                  ApiCallTimeoutException.create(apiCallTimeoutInMillis),
-                                                                  apiCallTimeoutInMillis);
-            context.apiCallTimeoutTracker(timeoutTracker);
-
-            execute(future);
-            return future;
+            return execute(future);
         }
 
-        public void execute(CompletableFuture<Response<OutputT>> future) throws Exception {
+        public CompletableFuture<Response<OutputT>> execute(CompletableFuture<Response<OutputT>> future) throws Exception {
             beforeExecute();
-            doExecute().handle((resp, err) -> handle(future, resp, err));
+            doExecute().whenComplete((resp, err) -> maybeRetry(future, resp, err));
+            return future;
         }
 
-        private Void handle(CompletableFuture<Response<OutputT>> future,
-                            Response<OutputT> resp,
-                            Throwable err) {
+        private void maybeRetry(CompletableFuture<Response<OutputT>> future,
```

Comment:
Hmm I don't really like that because it implies we always want to retry, but we only want to do that if it's a retryable exception, and not if it was a successful response.

Reply:
Fair enough, `retryIfNeeded` then? `maybeRetry` feels a bit like we are retrying based on mood to me ╮(╯_╰)╭